File: articles/personality-culture-and-the-problem-of-universality/r/analyze_personality_culture_universality.R
# Personality, Culture, and the Problem of Universality
# R workflow for synthetic cross-cultural personality analysis

root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_personality_culture_universality.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

traits <- c(
  "openness",
  "conscientiousness",
  "extraversion",
  "agreeableness",
  "neuroticism",
  "honesty_humility"
)

required <- c("participant_id", "culture_group", traits, "context_collectivism", "behavioral_manifestation")
missing <- setdiff(required, names(data))
if (length(missing) > 0) {
  stop(paste("Missing required columns:", paste(missing, collapse = ", ")))
}

group_summary <- aggregate(
  data[, c(traits, "context_collectivism", "behavioral_manifestation")],
  by = list(culture_group = data$culture_group),
  FUN = mean
)

group_n <- aggregate(
  data$participant_id,
  by = list(culture_group = data$culture_group),
  FUN = length
)
names(group_n)[2] <- "n"

group_summary <- merge(group_n, group_summary, by = "culture_group")
write.csv(group_summary, file.path(outputs, "r_group_summary.csv"), row.names = FALSE)

pooled_corr <- cor(data[, traits], use = "pairwise.complete.obs")
write.csv(pooled_corr, file.path(outputs, "r_pooled_trait_correlations.csv"))

upper_values <- function(mat) {
  mat[upper.tri(mat)]
}

pooled_values <- upper_values(pooled_corr)

rows <- data.frame(
  culture_group = character(),
  matrix_similarity_with_pooled = numeric(),
  n = integer()
)

for (group_name in unique(data$culture_group)) {
  group_df <- data[data$culture_group == group_name, ]
  group_corr <- cor(group_df[, traits], use = "pairwise.complete.obs")
  write.csv(group_corr, file.path(outputs, paste0("r_correlations_", group_name, ".csv")))

  similarity <- cor(pooled_values, upper_values(group_corr), use = "complete.obs")
  rows <- rbind(rows, data.frame(
    culture_group = group_name,
    matrix_similarity_with_pooled = round(similarity, 4),
    n = nrow(group_df)
  ))
}

write.csv(rows, file.path(outputs, "r_matrix_similarity.csv"), row.names = FALSE)

model <- lm(
  behavioral_manifestation ~ conscientiousness + agreeableness + honesty_humility + context_collectivism,
  data = data
)

capture.output(summary(model), file = file.path(outputs, "r_behavioral_model_summary.txt"))

cat("Wrote R outputs to:", outputs, "\n")

File: articles/personality-culture-and-the-problem-of-universality/python/analyze_personality_culture_universality.py
#!/usr/bin/env python3
"""Analyze synthetic cross-cultural personality data.

This script supports the article "Personality, Culture, and the Problem of Universality."
It demonstrates group summaries, within-group correlation matrices, and matrix-similarity
comparisons using synthetic data only.
"""

from __future__ import annotations

from pathlib import Path
import json
import numpy as np
import pandas as pd


ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "synthetic_personality_culture_universality.csv"
OUTPUTS = ROOT / "outputs"

TRAITS = [
    "openness",
    "conscientiousness",
    "extraversion",
    "agreeableness",
    "neuroticism",
    "honesty_humility",
]


def upper_triangle_values(matrix: pd.DataFrame) -> pd.Series:
    """Return upper-triangle values from a square correlation matrix."""
    mask = np.triu(np.ones(matrix.shape), k=1).astype(bool)
    return matrix.where(mask).stack()


def main() -> None:
    OUTPUTS.mkdir(parents=True, exist_ok=True)

    df = pd.read_csv(DATA_PATH)

    required = {"participant_id", "culture_group", *TRAITS, "context_collectivism", "behavioral_manifestation"}
    missing = required.difference(df.columns)
    if missing:
        raise ValueError(f"Missing required columns: {sorted(missing)}")

    group_summary = (
        df.groupby("culture_group")
        .agg(
            n=("participant_id", "count"),
            openness_mean=("openness", "mean"),
            conscientiousness_mean=("conscientiousness", "mean"),
            extraversion_mean=("extraversion", "mean"),
            agreeableness_mean=("agreeableness", "mean"),
            neuroticism_mean=("neuroticism", "mean"),
            honesty_humility_mean=("honesty_humility", "mean"),
            context_collectivism_mean=("context_collectivism", "mean"),
            behavioral_manifestation_mean=("behavioral_manifestation", "mean"),
        )
        .reset_index()
    )

    group_summary.to_csv(OUTPUTS / "python_group_summary.csv", index=False)

    pooled_corr = df[TRAITS].corr()
    pooled_corr.to_csv(OUTPUTS / "python_pooled_trait_correlations.csv")

    pooled_values = upper_triangle_values(pooled_corr)

    rows = []
    for group, group_df in df.groupby("culture_group"):
        corr = group_df[TRAITS].corr()
        corr.to_csv(OUTPUTS / f"python_correlations_{group}.csv")

        group_values = upper_triangle_values(corr)
        aligned = pd.concat([pooled_values, group_values], axis=1).dropna()
        aligned.columns = ["pooled", "group"]

        similarity = aligned["pooled"].corr(aligned["group"])
        rows.append(
            {
                "culture_group": group,
                "matrix_similarity_with_pooled": round(float(similarity), 4),
                "n": int(len(group_df)),
            }
        )

    replicability = pd.DataFrame(rows)
    replicability.to_csv(OUTPUTS / "python_matrix_similarity.csv", index=False)

    # Simple trait-context behavioral model using numpy least squares.
    # behavioral_manifestation ~ conscientiousness + agreeableness + honesty_humility + context_collectivism
    X = df[["conscientiousness", "agreeableness", "honesty_humility", "context_collectivism"]].to_numpy()
    X = np.column_stack([np.ones(len(X)), X])
    y = df["behavioral_manifestation"].to_numpy()
    beta, *_ = np.linalg.lstsq(X, y, rcond=None)

    model = {
        "intercept": float(beta[0]),
        "conscientiousness": float(beta[1]),
        "agreeableness": float(beta[2]),
        "honesty_humility": float(beta[3]),
        "context_collectivism": float(beta[4]),
        "note": "Synthetic demonstration only; not a real psychological model.",
    }

    with open(OUTPUTS / "python_behavioral_model.json", "w", encoding="utf-8") as f:
        json.dump(model, f, indent=2)

    print("Wrote Python outputs to:", OUTPUTS)


if __name__ == "__main__":
    main()
